pages/_app.js

```diff
@@ -7,10 +7,32 @@ import { nanoid } from "nanoid";
 export default function App({ Component, pageProps }) {
   const [shoppingItems, setShoppingItems] = useState(initialShoppingItems);
 
+  const shoppingItemsWithCategoryColor = shoppingItems.map((shoppingItem) => {
+    const category = categories.find(
+      (category) => category.name === shoppingItem.category
+    );
+
+    const backgroundColor = category ? category.color : "white";
+
+    return {
+      ...shoppingItem,
+      backgroundColor,
+    };
+  });
+
+  const purchasedItems = shoppingItemsWithCategoryColor.filter(
+    (shoppingItem) => shoppingItem.isPurchased
+  );
+
+  const unpurchasedItems = shoppingItemsWithCategoryColor.filter(
+    (shoppingItem) => shoppingItem.isPurchased === false
+  );
+
   function handleAddItem(newItem) {
     setShoppingItems([
```

Comment:
I recommend to be consistent when updating your state: In your `handleToggleIsPurchased` function, you are using `prevItems` during your state update. In all your other handler functions, where you update the `shoppingItems` state, you do not use `prevItems`.

React’s state updates are asynchronous. If you call `setShoppingItems` directly with `[newItem, ...shoppingItems]`, there’s a chance you’re referencing an outdated version of `shoppingItems`, especially if there are multiple state updates happening in quick succession or if other components are also updating the same state.
By using the callback `(prevItems) => [...]`, you are guaranteed to be working with the latest state, as React passes the most recent version of `shoppingItems` to `prevItems`.

```suggestion
 setShoppingItems((prevItems) => [
      {
        id: nanoid(),
        isPurchased: false,
        ...newItem,
      },
      ...prevItems,
    ]);
```

Reply:
The same applies to your `handleEditItem` and `handleDeleteItem` function:

```js
 function handleEditItem(id, newItem) {
    setShoppingItems((prevItems) =>
      prevItems.map((shoppingItem) => {
        if (shoppingItem.id === id) return { ...shoppingItem, ...newItem };
        return shoppingItem;
      })
    );
  }

  function handleDeleteItem(id) {
    setShoppingItems((prevItems) =>
      prevItems.filter((shoppingItem) => shoppingItem.id !== id)
    );
  }
```

Reply:
I have replaced prevItems with shoppingItems.

I now call setShoppingItems, handleEditItem and handleDeleteItem with callback (prevItems) rather than directly as before.